File: src/jzrlib/external/bsp/BSPNode.java
/*
 *  "JZRLib", Java Roguelike games development Library.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih (aka Alchemist, aka Norseman).
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package jzrlib.external.bsp;

import java.util.ArrayList;
import java.util.List;

/**
 *
 * @author Serg V. Zhdanovskih
 */
public final class BSPNode
{
    private static int lastId = 0;

    public final int id;
    public final BSPNode parent;
    public final int depth;
    public final int x1, y1, x2, y2;

    public BSPNode left;
    public BSPNode right;

    public LocationType nodeType;
    public SplitterType splitterType;

    public final List<NodeSide> sides;

    public BSPNode(BSPNode parent, int depth, int x1, int y1, int x2, int y2)
    {
        this.id = ++lastId;
        this.parent = parent;
        this.depth = depth;
        this.x1 = x1;
        this.y1 = y1;
        this.x2 = x2;
        this.y2 = y2;

        this.left = null;
        this.right = null;

        this.nodeType = LocationType.ltNone;
        this.splitterType = SplitterType.stNone;

        this.sides = new ArrayList<>();
    }

    public final int getWidth()
    {
        return (this.x2 - this.x1) + 1;
    }

    public final int getHeight()
    {
        return (this.y2 - this.y1) + 1;
    }

    public static void resetId()
    {
        lastId = 0;
    }
}

File: project/External/BSP/BSPNode.cs
/*
 *  "ZRLib", Roguelike games development Library.
 *  Copyright (C) 2015 by Serg V. Zhdanovskih.
 *
 *  This program is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

using System.Collections.Generic;

namespace ZRLib.External.BSP
{
    public enum SplitterType
    {
        stNone,
        stHorz,
        stVert
    }

    public enum LocationType
    {
        ltNone,
        ltInner,
        ltOuter
    }

    public sealed class BSPNode
    {
        private static int LastId = 0;

        public readonly int Id;
        public readonly BSPNode Parent;
        public readonly int Depth;
        public readonly int X1, Y1, X2, Y2;

        public BSPNode Left;
        public BSPNode Right;

        public LocationType NodeType;
        public SplitterType SplitterType;

        public readonly IList<NodeSide> Sides;

        public BSPNode(BSPNode parent, int depth, int x1, int y1, int x2, int y2)
        {
            Id = ++LastId;
            Parent = parent;
            Depth = depth;
            X1 = x1;
            Y1 = y1;
            X2 = x2;
            Y2 = y2;

            Left = null;
            Right = null;

            NodeType = LocationType.ltNone;
            SplitterType = SplitterType.stNone;

            Sides = new List<NodeSide>();
        }

        public int Width
        {
            get { return (X2 - X1) + 1; }
        }

        public int Height
        {
            get { return (Y2 - Y1) + 1; }
        }

        public static void ResetId()
        {
            LastId = 0;
        }
    }
}
